airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/HeartbeatTimeoutChaperone.java

```diff
@@ -0,0 +1,130 @@
+
+  public HeartbeatTimeoutChaperone(final HeartbeatMonitor heartbeatMonitor,
+                                   final Duration timeoutCheckDuration,
+                                   final FeatureFlagClient featureFlagClient,
+                                   final UUID workspaceId) {
+    this.timeoutCheckDuration = timeoutCheckDuration;
+
+    LOGGER.info("Starting source heartbeat check. Will check every {} minutes.", timeoutCheckDuration.toMinutes());
+
+    this.heartbeatMonitor = heartbeatMonitor;
+    this.featureFlagClient = featureFlagClient;
+    this.workspaceId = workspaceId;
+  }
+
+  public Runnable runWithHeartbeatThread(final Runnable runnable, final ExecutorService executorService) {
+    return () -> {
+      final CompletableFuture<Void> runnableFuture = CompletableFuture.runAsync(runnable, executorService);
+      final CompletableFuture<Void> heartbeatFuture = CompletableFuture.runAsync(this::monitor, executorService);
+
+      try {
+        CompletableFuture.anyOf(runnableFuture, heartbeatFuture).get();
+      } catch (final InterruptedException e) {
+        LOGGER.error("Heartbeat chaperone thread was interrupted.", e);
+        return;
+      } catch (final ExecutionException e) {
+        // this should check explicitly for source and destination exceptions
+        if (e.getCause() instanceof RuntimeException) {
+          throw (RuntimeException) e.getCause();
+        } else {
+          throw new RuntimeException(e);
+        }
+      }
+
+      LOGGER.info("thread status... heartbeat thread: {} , replication thread: {}", heartbeatFuture.isDone(), runnableFuture.isDone());
+
+      boolean heartbeatTimedOut = false;
+      if (heartbeatFuture.isDone() && !runnableFuture.isDone()) {
+        heartbeatTimedOut = true;
+        runnableFuture.cancel(true);
+      }
+
+      if (runnableFuture.isDone() && !heartbeatFuture.isDone()) {
+        heartbeatFuture.cancel(true);
+      }
+
+      if (heartbeatTimedOut) {
+        throw new HeartbeatTimeoutException(String.format("Heartbeat has stopped. Heartbeat freshness threshold: %s Actual heartbeat age: %s",
+            heartbeatMonitor.getHeartbeatFreshnessThreshold(),
+            heartbeatMonitor.getTimeSinceLastBeat().orElse(null)));
+      }
+    };
+  }
+
+  @SuppressWarnings("BusyWait")
+  private void monitor() {
+    while (true) {
+      try {
+        sleep(timeoutCheckDuration.toMillis());
+      } catch (final InterruptedException e) {
+        LOGGER.info("Heartbeat thread has been interrupted (this is expected; the heartbeat was healthy the whole time).");
+        return;
+      }
+
+      // if not beating, return. otherwise, if it is beating or heartbeat hasn't started, continue.
+      if (!heartbeatMonitor.isBeating().orElse(true)) {
+        if (!hasFailed) {
+          LOGGER.error("Source has stopped heart beating.");
+          if (featureFlagClient.enabled(ShouldFailSyncIfHeartbeatFailure.INSTANCE, new Workspace(workspaceId))) {
```

Comment:
This looks a bit confusing, I'd expect the heartbeatMonitor to always fail the heartbeats fails. I'd have the feature flag in the caller (`runWithHeartbeatThread`?) to decide if we want to abort or not.

Reply:
the goal is to be able to monitor the potential forced failure. I just want to feature flag the action of making the connection to fail. The monitoring is added in another PR which is currently in draft: https://github.com/airbytehq/airbyte/pull/22584/files#diff-64e01615c3f046cc79b2df9bd1e55b088345ccb4ad23290d57aa466317d2ec2eR119 Does it explain why the feature flag is not on the call to monitor?